### ISSUE TITLE
Possible bug for AsyncCircuitBreakerPolicy when circuit is in half-open state and exception handled by the policy is raised after exception not handled by the policy.

### ISSUE THREAD

**Issue by EnricoMassoneDeltatre** [App-vNext/Polly#717]
Hi,

I observed something unexpected with the Polly's implementation of circuit breaker pattern while I was writing some unit tests on one of my classes making use of Polly. I suspect this is a bug on Polly side. 

Based on my understanding, given a policy of type `AsyncCircuitBreakerPolicy` when the circuit is in the **half-open state**, then the circuit **is never expected to throw an exception of type `BrokenCircuitException`**. Instead, it is expected to always rethrow the policy delegate exception, whether it is handled by the policy itself or not. 

This is exactly the behavior documented [here](https://github.com/App-vNext/Polly/wiki/Circuit-Breaker#putting-it-all-together-).

We are experiencing the following unexpected behavior. 

We have a policy of type `AsyncCircuitBreakerPolicy` and the circuit is in the **half-open** state.
We execute the following two actions:

1. execution of a call to the policy delegate so that an exception **not handled** by the policy is raised. The expectation is that the exception is rethrown as is by the policy and that the circuit stays in the **half-open** state. This expectation is met.

2. execution of a call to the policy delegate so that an exception **handled** by the policy is raised. The expectation is that the exception is rethrown as is by the policy and that the circuit changes its state from **half-open** to **open**. This expectation is **not** met. Instead, an exception of type 
`BrokenCircuitException` is thrown by the policy. We **never** expect that a circuit in the **half-open** state throws an exception of type `BrokenCircuitException`.
 
Interestingly, if a delay **long enough** is inserted between steps 1 and 2, then the policy behavior is the expected one. We also observed that **the minimum delay length required to get the expected behavior is the duration of break passed when the policy was instantiated**.
 
This is weird: it's the kind of issue that you experience when you have asynchronous methods in your code and you forget to put an `await` keyword somewhere. 

You can use the following program to reproduce the issue we are experiencing.

If you run it exactly the way it is written here, it will break with an unhandled exception of type `BrokenCircuitException`.

```
using System;
using System.Runtime.Serialization;
using System.Threading.Tasks;
using Polly;

namespace Polly.CircuitBreaker.Demo
{
  class Program
  {
    static async Task Main(string[] args)
    {
      AsyncCircuitBreakerPolicy policy = Policy
        .Handle<Exception>(ex => ex is TestException)
        .CircuitBreakerAsync(2, TimeSpan.FromSeconds(3));

      try
      {
        await policy.ExecuteAsync(() => WaitAndThrows(new TestException())).ConfigureAwait(false);
      }
      catch (TestException)
      {
        Console.WriteLine($"TestException handled. State: {policy.CircuitState}"); // expected state: closed
      }

      try
      {
        await policy.ExecuteAsync(() => WaitAndThrows(new TestException())).ConfigureAwait(false);
      }
      catch (TestException)
      {
        Console.WriteLine($"TestException handled. State: {policy.CircuitState}"); // expected state: open
      }

      Console.WriteLine($"State: {policy.CircuitState}"); // expected state open

      try
      {
        await policy.ExecuteAsync(() => WaitAndThrows(new Exception())).ConfigureAwait(false);
      }
      catch (BrokenCircuitException)
      {
        Console.WriteLine($"BrokenCircuitException handled. State: {policy.CircuitState}"); // expected state open
      }

      try
      {
        await policy.ExecuteAsync(() => WaitAndThrows(new TestException())).ConfigureAwait(false);
      }
      catch (BrokenCircuitException)
      {
        Console.WriteLine($"BrokenCircuitException handled. State: {policy.CircuitState}"); // expected state open
      }

      await Task.Delay(TimeSpan.FromSeconds(3)).ConfigureAwait(false);

      Console.WriteLine($"State: {policy.CircuitState}"); // expected state half open

      try
      {
        await policy.ExecuteAsync(() => WaitAndThrows(new Exception())).ConfigureAwait(false);
      }
      catch (Exception)
      {
        Console.WriteLine($"Exception handled. State: {policy.CircuitState}"); // expected state half open
      }

      Console.WriteLine($"State: {policy.CircuitState}"); // expected state half open

      await Task.Delay(TimeSpan.FromSeconds(1)).ConfigureAwait(false); // if you change this to 3 seconds it works

      try
      {
        Console.WriteLine($"State: {policy.CircuitState}"); // expected state half open
        // this call is expected to throw an exception of type TestException but, instead, an exception of type BrokenCircuitException is thrown
        await policy.ExecuteAsync(() => WaitAndThrows(new TestException())).ConfigureAwait(false);
      }
      catch (TestException)
      {
        Console.WriteLine($"TestException handled. State: {policy.CircuitState}");
      }

      Console.WriteLine($"State: {policy.CircuitState}");
    }

    static async Task Wait()
    {
      await Task.Delay(10).ConfigureAwait(false);
    }

    static async Task WaitAndThrows(Exception exception)
    {
      await Task.Delay(10).ConfigureAwait(false);
      throw exception;
    }

    [Serializable]
    public class TestException : Exception
    {
      public TestException()
      {
      }

      public TestException(string message) : base(message)
      {
      }

      public TestException(string message, Exception innerException) : base(message, innerException)
      {
      }

      protected TestException(SerializationInfo info, StreamingContext context) : base(info, context)
      {
      }
    }
  }
}
```

By changing the last call to `Task.Delay` as suggested in the comment, the program works as expected without any unhandled exception.

This is my csproj:

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp3.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Polly" Version="7.1.1" />
  </ItemGroup>

</Project>
```

I'm running this code on a windows 10 machine, from within Visual Studio 2019.

**Comment by reisenberger** [comment]
Apologies for brief reply: I am away. This looks to me like expected behaviour. Duplicate of #311 .

**Comment by EnricoMassoneDeltatre** [comment]
> Apologies for brief reply: I am away. This looks to me like expected behaviour. Duplicate of #311 .

Hi and thanks for replying.

I wasn't aware of the "single trial per break duration rule" when the circuit is in the **half-open** state. I actually read about it in the documentation, but I didn't fully understand all the implications of that. 

So, just to summarize, we can state the following:

- when the circuit is in the **half-open** state only one attempt to execute the delegate is permitted during a time frame equals to the configured break duration.
- if this attempt succeeds without throwing any exception, then the circuit transitions to the **closed** state. 
- if an exception is thrown during the attempt, **the exception is rethrown by the policy whether or not it is handled by the policy itself**. If the exception is one of the handled exceptions, then the circuit moves to the **open** state. Otherwise, it remains in the **half-open** state. 
 - any other attempt to execute the delegate will be immediately rejected by throwing the `BrokenCircuitException`, without executing the delegate (fail fast approach, the same observed when the circuit is in the **open** state). 
 - if the circuit moves to the **open** state, it will remain open for the entire break duration (as usual). If it remains in the **half-open** state, **other attempts to execute the delegate will be allowed only after a delay equals to the break duration**. Only one attempt is allowed per break duration.

Is all of this correct ?

Thanks!

**Comment by reisenberger** [comment]
@EnricoMassoneDeltatre Yes, that's correct.
